User can renew a book

margaretoti opened the pull request:
## UPDATE 05.02.2017
Users can renew a book as many times as they like from the profile page. Creating a checkout, renewing a book, and returning a book are now handled by ruby service objects. When a user renews a book, all prior fines are displayed. In future, users will only be able to renew books a maximum of 2 times before having to return the books to the library. 

## Implementation Details 
### Renewals
In prior commits, when a user renewed a book from her profile, all associated fines would disappear. This is because the book (technically the checkout) is no longer overdue.

To avoid this bug where accumulated fines are erased when books are renewed, the underlying renewal mechanism has changed. Previously when a user clicked on the `Renew` link, the existing checkout was updated to reflect the new `due_on` date and the `fine` was calculated and stored on the checkout. **Now** when the user clicks `Renew`, she is directed to a new `renewals route` and the followings actions occur (also see the section on Database Changes ):
1. The `fine` on the original checkout is calculated and stored on the original checkout. `closed_at` is marked with the current time. 
2. A new checkout is created with `fine` set to 0 and `closed_at` marked as`nil`.

Thus, fines live on the checkout that they were accrued on. To determine the total amount that user owes for a specific book, all of the checkouts (with the same `book_id` and `user_id`) are summed.

The total fines owed across all books (technically all checkouts) remains a calculated “running” value.

_Advantages_
The same checkout is not continually being updated to reflect the changes the user has made. This makes it easier to:
* Store and display fines and fine history
* Store and display checkout history.
* Calculate the number of times a user has renewed a book (a future story).

### Database changes
The database columns `closed_at` and `fine_cents` ([money-rails gem](https://github.com/RubyMoney/money-rails)) have now been added to the `Checkouts` table. The unique constraint on book has been adjusted a partial unique id to reflect that more than one checkout can exist with the same `book_id`.

New structure of Checkouts table
`id` | `book_id` | `user_id` | `fine_cents` | `closed_at`

### Displaying Available Books
I used a LEFT JOIN because `Book.joins(:checkouts).where.not(checkouts: {closed_at: nil })` displayed books as available if ANY of the checkouts associated with that `book_id` and `user_id` where marked as closed. I need books to be listed as available only if EVERY checkout is marked
as closed.

Details: The inner query finds all checkouts yet to be returned. Then I join them with all books. Using a LEFT JOIN means that if a book is not checked out, it will still be included in the results table, just with `nil` values for the checkout.

### Returns
When a book is returned, any additional overdue fines are calculated and stored on the last checkout (which should be the only open checkout for that book). The checkout is marked as closed. 

In prior branches, all checkouts with the same `book_id` and `user_id` where destroyed when the user clicked `Return`. This approach made it difficult to track, store, and calculate fines on books that a user had returned.

### New Additions 
Some tools, features, and patterns used in this branch include:
- Service objects
- Renewals route and Renewals controller
- Money-rails gem
- SQL Left Joins 

### Up Next
1. Refactor tests
2. Make UI clearer for the user
- Add original checkout date to table in user profile
- Add renewal times and limit them to 2
3. Users can pay fines
4. Users can pay partial fines

margaretoti committed:
Add due_on column to checkouts table

due_on will now handle due dates. The method due date in the checkouts
model is no longer used as due dates are now persisted into the
database.

margaretoti committed:
Model methods handle borrowing and renewing books

margaretoti committed:
Use Ruby service objects

Service objects now handle checking out a book, renewing a book, and
returning a book.

Success messages have also been added.

margaretoti commented on app/controllers/checkouts_controller.rb:
```diff
@@ -3,13 +3,26 @@ class CheckoutsController < ApplicationController
     @book = Book.find(params[:book_id])
-    checkout = Checkout.create(book: @book, user: current_user)
+    if CheckOutBook.process(book: @book, user: current_user)
+      flash[:success] = t('.successful_message')
```
@derekprior, Is it a better idea to have the Ruby service object handle success and error/failure? The controller was so sleek until I added the `if else` clause.

derekprior replied:
`redirect_to` accepts options for setting the flash. By default, it supports the `notice` and `alert` keys. We're using `success` and `error` here, so we can register those using `add_flash_types` in `ApplicationController`: http://api.rubyonrails.org/classes/ActionController/Flash/ClassMethods.html

Once that's done setting the flash and redirecting can be a one liner everywhere, which is nice.

derekprior commented on app/controllers/checkouts_controller.rb:
```diff
@@ -3,13 +3,26 @@ class CheckoutsController < ApplicationController
+      flash[:success] = t('.successful_message')
+      redirect_to root_path
+    else
+      flash[:error] = t('.unsuccessful_message')
```
This action will attempt to render the `create` view, by default. Should it redirect? Or `render :new` (if there's a form for this action -- I can't remember)?

margaretoti replied:
hi @derekprior, the checkouts controller does not have any associated views. All creating a checkout (i.e. checking out a book from the library does) does is create a record in the database, mark the book as unavailable, and display that book on your profile where it can be renewed or returned.

margaretoti replied:
Redirecting here simply serves as a page refresh so the user can see the updated library catalog. The book they just checked out should now be listed as unavailable.

margaretoti replied:
Redirecting here simply serves as a page refresh so the user can see the updated library catalog. The book they just checked out should now be listed as unavailable.

derekprior replied:
You're not redirecting in this branch though. So I think if you did get an error in the checkout process somehow (maybe someone else checked the book out right before you did), I think rails would raise an error because it will attempt to render the :create view by default. Same with the update action below, actually.

I think you need the redirect_to call here too.

margaretoti replied:
I understand now. Because I did not specify a redirect in the else clause, it will try to render create, which will fail. A failure I will never catch because I don't know how test the scenario where someone else checked the book out right before you did. I removed both `else` clauses and added `create!` and `update!` in `RenewBook` and `CheckOutBook` services respectively instead.

derekprior reviewed (commented):
This looks great. The service objects are really simple and I think it was a good direction. As you tackle more complex things you'll probably be glad to have tried them out! Good work.

derekprior reviewed (commented):
Looking good!

margaretoti committed:
Renewing a book creates a new checkout

This PR accomplishes 3 things:

1. Renewing a book creates a new checkout
2. Fines display when overdue books are renewed
3. Implement a SQL LEFT JOIN to correctly display available books in the
library catalog.

To avoid the bug where accumulated fines are erased when a book is
renewed, the existing checkout is no longer updated to reflect the new
`due_on ` date. Instead, a new checkout is created. This new checkout
stores the total fine across all checkouts/renewals with the same
`book_id` and `user_id`. It also stores the new `due_on` date.

The total fines owed across all books (technically checkouts) remains a
calculated/ “running” value.

New structure of Checkouts table
`id` | `book_id` | `user_id` | `fine_cents` | `closed_at`

The unique constraint on `book_id` has been removed. Column `fine_cents`
is provided by money-rails gem.

Returns
When a book is returned, any additional overdue fines are calculated and
stored on the checkout. The checkout is marked as closed.

Previously when a book was returned, all checkouts containing the same
`book_id` and `user_id` were destroyed. Thus the mechanism for displaying
available and unavailable books in the library catalog based on whether
a checkout existed worked. That logic has now been replaced with a SQL
LEFT JOIN for available books as every checkout associated with a book
must be closed for the book to be available to the public.

margaretoti committed:
Restructure overdue fine storage

Users can renew a book as many times as they like from the profile page.
Creating a checkout, renewing a book, and returning a book are now
handled by ruby service objects. When a user renews a book, all prior
fines are displayed. In future, users will only be able to renew books a
maximum of 2 times before having to return the books to the library.

Implementation Details

Renewals
In prior commits, when a user renewed a book from her profile, all
associated fines would disappear. This is because the book (technically
the checkout) is no longer overdue.

To avoid this bug where accumulated fines are erased when book are
renewed, the underlying renewal mechanism has changed. Previously when a
user clicked on the `Renew` link, the existing checkout was updated to
reflect the new `due_on` date and the `fine` was calculated and stored
on the checkout. **Now**, when the user clicks on `Renew`, she is
directed to a new `renewals route` and the following actions occur (also
see section on the Database Changes):
1. The `fine` on the original checkout is calculated and stored on the
original checkout. `closed_at` is marked with the current time.
2. A new checkout is created with `fine` set to 0 and `closed_at` marked
as `nil`.

Thus, fines live on the checkout that they were accrued on. To determine
the total amount that the user owes for a specific book, all of the
checkouts (with the same `book_id` and `user_id`) are summed.

The total fines owed across all books (technically all checkouts)
remains a calculated "running" value.

Advantages
The same checkout is not continually being updated to reflect the
changes the user has made. This makes it easier to:
* Store and display fines and fine history
* Store and display checkout history
* Calculate the number of times a user has renewed a book (a future
story).

Database Changes
The database columns `closed_at` and `fine_cents` (money-rails gem) have
now been added to the `Checkouts` table. The unique constraint on the
book has been adjusted to a partial unique id to reflect that more than
one checkout can exist with the same `book_id`.

New Structure of Checkouts Table
`id` | `book_id` | `user_id` | `fine_cents` | `closed_at`

Displaying Available Books
I used a LEFT JOIN because `Book.joins(:checkouts).where.not(checkouts:
{ closed_at: nil })` displayed books as available if ANY of the
checkouts associated with that `book_id` and `user_id` where marked as
closed. I need books to be listed as available only if EVERY checkout is
marked as closed.

Details: The inner query finds all checkouts yet to be returned. Then I
join them with all books. Using a LEFT JOIN means that if a book is not
checked out, it will still be included in the results table, just with
`nil` values for the checkout.

Returns
When a book is returned, any additional overdue fines are calculated and
stored on the last checkout (which should be the only open checkout for
that book). The checkout is marked as closed.

In prior branches, all checkouts with the same `book_id` and `user_id`
where destroyed when the user clicked `Return`. This approach made it
difficult to track, store, and calculate fines on books that a user had
returned.

New Additions
Some tools, features, and patterns used in this branch include:
* Service objects
* Renewals route and Renewals controller
* Money-rails gem
* SQL Left Joins

Up Next
1. Refactor tests
2. Make UI clearer for the user
- Add original checkout date to table in user profile
- Add renewal times and limit them to 2
3. Users can pay fines
4. Users can pay partial fines

margaretoti commented on app/models/checkout.rb:
```diff
@@ -1,34 +1,43 @@
-  end
+  monetize :fine_cents
 
-  def fine
```
@paulcsmith, Should I abstract all methods related to fines to a `Fine model`? There are fine methods is `checkout.rb` and `user.rb`.

paulcsmith replied:
I would probably create a `CalculateFine` class that does this. I imagine it would be used a bit like this:

```ruby
CalculateFine.for(checkout)
CalculateFine.including_related_checkout_for(checkout)
```

margaretoti commented on spec/features/user_can_renew_book_spec.rb:
```diff
@@ -0,0 +1,94 @@
+
+    checkout.reload
+
+    # one_day_overdue = 1
```
@paulcsmith, This commented out method of calculating the fine matches the old approach where the fine was "bubbled up" from each checkout and only ever lived on one checkout. I am thinking of a way to duplicate this test so it fits the current approach of having each checkout reflect the fine associated with it's checkout period.

margaretoti commented on spec/features/user_sees_fine_amounts_spec.rb:
```diff
@@ -16,7 +16,7 @@
     expect(page).to have_content(t('profiles.show.fine_total'))
     expect(page).to have_total_fines_of('$0.10')
-    expect(page).to not_have_any_fines_of_zero_dollars
+    # expect(page).to not_have_any_fines_of_zero_dollars
```
@paulcsmith, I now allow fines of $0.00 to show in the UI. I need to think about whether I want this to be the case.

margaretoti commented on spec/models/checkout_spec.rb:
```diff
@@ -4,9 +4,10 @@
 
     subject { FactoryGirl.build(:checkout) }
-    it { should validate_uniqueness_of(:book) }
+    # it { should validate_uniqueness_of(:book) }
```
@paulcsmith, Partial index will be added shortly.

margaretoti commented on app/models/book.rb:
```diff
@@ -5,11 +5,22 @@ class Book < ApplicationRecord
+    # displayed checkouts where ANY checkout associated with the book
+    # and user would be listed as unavailable. I need books to listed as
+    # available when EVERY checkout is closed. 
+    Book.joins(
```
@paulcsmith, See comment above for explanation as to `SQL join`.

paulcsmith replied:
Thanks for the comment. I would probably add that to the commit message rather than in the code itself. More details here: https://robots.thoughtbot.com/letting-your-code-speak-for-itself

Also you may be able to use these new methods in Rails 5 instead of dropping down to raw SQL: http://blog.bigbinary.com/2016/03/24/support-for-left-outer-joins-in-rails-5.html

margaretoti replied:
@paulcsmith, I used a `LEFT JOIN` because `Book.joins(:checkouts).where.not(checkouts: {closed_at: nil })` displayed books as available if ANY of the checkouts associated with that `book_id` and `user_id` where marked as closed. I need books to be listed as available only if EVERY checkout is marked as closed. 

Details: The inner query finds all checkouts yet to be returned.  Then I join them with all books. Using a `LEFT JOIN` means that if a book is not checked out, it will still be included in the results table, just with nil values for the checkout.

Any advice on how to make the SQL match a Rails construct would be great.

paulcsmith commented on app/controllers/application_controller.rb:
```diff
@@ -3,6 +3,8 @@ class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
   after_action :ensure_authorization_checked
 
+  add_flash_types :success, :error 
```
I didn't know about this. Very cool!

margaretoti replied:
Derek shared this nice tip with me.

paulcsmith commented on app/controllers/checkouts_controller.rb:
```diff
@@ -3,14 +3,18 @@ class CheckoutsController < ApplicationController
+  end
 
-    redirect_to root_path
+  def update
```
I would probably put this in a separate resource like `:renewals` and do it on `create` so you're "creating a renewal" , rather than "updating a checkout" (which is actually renewing it)

paulcsmith commented on app/controllers/profiles_controller.rb:
```diff
@@ -2,8 +2,7 @@ class ProfilesController < ApplicationController
   def show
-    @checkouts = current_user.checkouts.includes(:book)
-
+    @checkouts = current_user.checkouts.includes(:book).where(closed_at: nil)
```
I would extract a scope for the `where(closed_at: nil)`. Something like `closed` maybe?

paulcsmith commented on app/models/checkout.rb:
```diff
@@ -1,34 +1,43 @@
 class Checkout < ApplicationRecord
   CHECKOUT_PERIOD_IN_DAYS = 14
-  FINE_AMOUNT_PER_DAY_IN_CENTS = 0.10
+  FINE_AMOUNT_PER_DAY_IN_CENTS = 10
```
Good name!

paulcsmith commented on app/models/checkout.rb:
```diff
@@ -1,34 +1,43 @@
   def overdue?
-    number_of_days_overdue = days_overdue
-    number_of_days_overdue > 0 ? true : false
+    if closed_at.present? # checkout is not overdue if it has been closed
```
As I mentioned in the blog post I linked to in a previous comment in this review, comments might not be the best way to do this. Instead you can extract a method for this `if closed?`

paulcsmith commented on app/models/checkout.rb:
```diff
@@ -1,34 +1,43 @@
+    if closed_at.present? # checkout is not overdue if it has been closed
+      false
+    else
+      days_overdue > 0 ? true : false
```
When you do comparison it already returns `true` or `false`, so you can remove the ternary

margaretoti replied:
Good call. I will definitely change that.

paulcsmith commented on app/models/checkout.rb:
```diff
@@ -1,34 +1,43 @@
 
   def days_overdue
-    (Time.zone.now - due_date).to_i/1.day
+    (Time.zone.now - self.due_on).to_i/1.day
```
I don't think you need `self` for this. Also it's a bit hard to see the division here. I'd put spaces around the `/`

paulcsmith replied:
Also Rails gives you `Time.current` which I think will do the same thing. Also, I'm not sure you need to divide by 1.day

paulcsmith commented on app/models/user.rb:
```diff
@@ -7,7 +7,7 @@ class User < ApplicationRecord
 
   def total_fines
-    total = checkouts.inject(0) { |total, checkout| total + checkout.fine }
+    total = checkouts.inject(0) { |total, checkout| total + checkout.calculate_fine }
```
I think you can use `sum` for this. Note that I'm talking about Ruby's sum and not Rails. The Rails one is for ActiveRecord relations

margaretoti replied:
That is the mistake that I made. oops, I tried using Rails `sum` and struggled.

paulcsmith commented on app/services/check_out_book.rb:
```diff
@@ -0,0 +1,19 @@
+  end
+
+  def process
+    due_date = DateTime.now + Checkout::CHECKOUT_PERIOD_IN_DAYS.days
```
I'd use `Time.current` here. The guideline I use is to use `Time` unless I *need* to use DateTime for something. I can't remember the last time I've needed DateTime though so Time is pretty much the go to class to use

paulcsmith commented on app/services/renew_book.rb:
```diff
@@ -0,0 +1,24 @@
+    book = Book.find(@checkout.book_id)
+    due_date = @checkout.due_on + Checkout::CHECKOUT_PERIOD_IN_DAYS.days
+    book_fine = @checkout.calculate_fine
+    closed_at = DateTime.now
```
I wouldn't put this in a var. It's extra stuff to read, but it doesn't make it easier to understand over doing `closed_at: Time.now`

paulcsmith commented on app/services/renew_book.rb:
```diff
@@ -0,0 +1,24 @@
+  def process
+    book = Book.find(@checkout.book_id)
+    due_date = @checkout.due_on + Checkout::CHECKOUT_PERIOD_IN_DAYS.days
+    book_fine = @checkout.calculate_fine
```
I'd put this, `book` and `due_date` in private methods. It will make it easier to see exactly what this method does

paulcsmith commented on app/services/return_book.rb:
```diff
@@ -0,0 +1,21 @@
+
+  def process
+    checkout = Checkout.where(book_id: @book_id, user_id: @user_id).last
+    book_fine = checkout.calculate_fine
```
You may want to extract private methods for these as well, so you can do:

```ruby
def process
  last_checkout.update!(fine: book_fine, closed_at: Time.current)
end
```

paulcsmith commented on app/services/return_book.rb:
```diff
@@ -0,0 +1,21 @@
+  end
+
+  def process
+    checkout = Checkout.where(book_id: @book_id, user_id: @user_id).last
```
I would suggest adding an explicit order to this so you don't accidentally call `last` and you get the wrong one

paulcsmith commented on db/migrate/20170424182702_remove_unique_constraint_on_book_in_checkouts.rb:
```diff
@@ -0,0 +1,6 @@
+class RemoveUniqueConstraintOnBookInCheckouts < ActiveRecord::Migration[5.0]
+  def change
+    remove_index :checkouts, :book_id
+    add_index :checkouts, :book_id
```
I think this is the part you said you were working on, right?

paulcsmith commented on spec/factories.rb:
```diff
@@ -10,5 +10,6 @@
   factory :checkout do
     user
     book
+    due_on DateTime.now + Checkout::CHECKOUT_PERIOD_IN_DAYS.days
```
Time.current

paulcsmith commented on spec/features/user_can_renew_book_spec.rb:
```diff
@@ -0,0 +1,94 @@
+    expect(page).to have_content(t('checkouts.update.successful_message'))
+  end
+
+  scenario 'prior fines carry over when an overdue book is renewed' do
```
I think if you say "carry over"  that implies that fines are moving. Instead I would consider saying "prior fines are included..."

paulcsmith reviewed (commented):
I think either you and I or @derekprior should pair on some of the tests to go over some different strategies for refactoring it. I left some comments on the other parts though. Maybe we can pair for a bit on Friday or Monday morning

margaretoti committed:
Add fine calculator objects

Added three fine calculator objects. 1. Checkout fine calculator object
to calculate the fine for a single checkout. 2. Book fine calculator
object that calculates the fine for a single book that the user checked
out. It encompasses all checkouts with the `current user_id` and the
`book_id`  3. User fine calculator object that calculates the total
number of fines a user has across all books the user has ever checked
out.

margaretoti committed:
Refactor tests

margaretoti committed:
Add partial unique index to book_id

Added a partial unique index to book_id in the checkouts table, ensuring
that one checkout per book is open. Now the UI alone is not
responsible for ensuring that a book is not checked out twice. A
possible, though unlikely scenario is that the same book is checked out
by two users at the same time (via the UI). This change helps to prevent
that occurrence.

margaretoti committed:
Small additions

1. Add open scope methods for checkouts
2. Add renewals controller
3. Add ApplicationRecord Transaction

margaretoti commented on app/models/book.rb:
```diff
@@ -2,14 +2,20 @@ class Book < ApplicationRecord
-  def self.unavailable
-    includes(:checkouts).where.not(checkouts: { id: nil })
+  def self.available
+    Book.joins(
```
I was not quite sure how to convert this to Rails 5. The things I tried did not work.

derekprior replied:
I think this is fine for this PR. I wouldn't want to chase it much further here given the size of this change already. I have some ideas on how you could refactor this to use active record associations and methods exclusively, but nopthing I'm really positive would work. Maybe we can pair on this later after this change is in.

margaretoti replied:
Thanks @derekprior, Let us definitely pair on this.

derekprior commented on app/models/user.rb:
```diff
@@ -6,8 +6,8 @@ class User < ApplicationRecord
-    total = checkouts.inject(0) { |total, checkout| total + checkout.fine }
-    total = sprintf('%.2f', total)
-  end
+  # def total_fines
```
Did you mean to leave this commented out? Typically we'd delete rather than comment out code. If we need it at a later date, we can awlays get it back from git history.

derekprior commented on app/services/book_fine_calculator.rb:
```diff
@@ -0,0 +1,22 @@
+  end
+
+  def including_related_checkouts
+    sum = checkouts_for_book
```
I don't think you need to assign the local variable `sum` here. You can return the result of the method call directly without the assignment.

Good use of inject here.

derekprior reviewed (commented):
This came along nicely. I like the use of the relatively small service objects to do the lifting here. I think there are areas for future improvement, but nothing incredibly pressing that should prevent you from moving forward here. 

Perfect is the enemy of good and that's especially true when changes start to get large (and this isn't too large but it's getting up there). This is good.

derekprior commented:
Oh, also glad to see the use of the money gem. I was going to suggest that as a future refactor after this change anyway.
